src/main/java/io/confluent/copycat/hdfs/WAL.java

```diff
@@ -0,0 +1,154 @@
+  private final String leaseException =
+      "org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException";
+
+  public WAL(Configuration conf, String url, String topicsDir, TopicPartition topicPart) {
+    try {
+      this.conf = conf;
+      this.topicsDir = topicsDir;
+      this.topicPart = topicPart;
+      this.url = url;
+      fs = FileSystem.get(URI.create(url), conf);
+      logFile = FileUtils.logFileName(url, topicsDir, topicPart);
+    } catch (IOException e) {
+      throw new CopycatRuntimeException(e);
+    }
+  }
+
+  public void append(String tempFile, String committedFile) throws InterruptedException {
+    while (true) {
+      try {
+        if (writer == null) {
+          writer = SequenceFile.createWriter(conf,
+                                             Writer.file(new Path(logFile)),
+                                             Writer.keyClass(Text.class),
+                                             Writer.valueClass(Text.class));
+        }
+        writer.append(new Text(tempFile), new Text(committedFile));
+        writer.hsync();
+        break;
+      } catch (RemoteException e) {
+        if (e.getClassName().equals(leaseException)) {
+          sleepIntervalMills = sleepIntervalMills * 2;
+          log.info("Cannot acquire lease on WAL {}", logFile);
+          Thread.sleep(sleepIntervalMills);
+        } else {
+          throw new CopycatRuntimeException(e);
+        }
+      } catch (IOException e) {
+        throw new CopycatRuntimeException("Error creating writer for log file " + logFile, e);
+      }
+    }
+  }
+
+  public void apply() {
+    try {
+      if (!fs.exists(new Path(logFile))) {
+        return;
+      }
+      if (reader == null) {
+        reader = new SequenceFile.Reader(conf, Reader.file(new Path(logFile)));
+      }
+      Text key = new Text();
+      Text value = new Text();
+      while (reader.next(key, value)) {
+        String tempFile = key.toString();
+        String committedFile = value.toString();
+
+        String[] offsets = committedFile.split("_");
+        final long startOffset = Integer.parseInt(offsets[0]);
+        Path path = new Path(FileUtils.directoryName(url, topicsDir, topicPart));
+        Path tempPath = new Path(tempFile);
+
+        if (fs.exists(tempPath)) {
+          FileStatus[] status = fs.listStatus(path, new PathFilter() {
+            @Override
+            public boolean accept(Path path) {
+              String filename = path.getName();
+              if (filename.equals("log")) {
+                return false;
+              }
+              String[] parts = path.getName().split("_");
+              if (parts.length != 2) {
+                return false;
+              }
+              if (parts[1].equals("tmp")) {
+                return false;
+              } else {
+                return Integer.parseInt(parts[0]) == startOffset;
+              }
+            }
+          });
+          if (status == null || status.length == 0) {
```

Comment:
I'm confused about what this block is doing. If an entry is added to the WAL, why would we ever delete it? Once the WAL entry is written, we should *always* deliver it, and retry until we succeed.